File: solutions/kotlin/sum-of-multiples/2/src/main/kotlin/SumOfMultiples.kt
object SumOfMultiples {

    fun sum(factors: Set<Int>, limit: Int): Int {
        var divisors = factors.filter{ it > 0 }
        var sum = 0
        for(n in 1 until limit) {
            if(divisors.any{ n % it == 0 }) sum += n
        }

        return sum
    }
}

File: solutions/zig/sum-of-multiples/2/sum_of_multiples.zig
const std = @import("std");
const mem = std.mem;

pub fn sum(allocator: mem.Allocator, factors: []const u32, limit: u32) !u64 {
    var set = std.AutoHashMap(u64, bool).init(allocator);
    defer set.deinit();

    var total: u64 = 0;

    for (factors) |f| {
        if (f <= 0) continue;
        var mult = f;
        while (mult < limit) : (mult += f) {
            if (!set.contains(mult)) {
                try set.put(mult, true);
                total += mult;
            }
        }
    }

    return total;
}
